subsys/bluetooth/host/l2cap.c

```diff
@@ -1256,6 +1252,25 @@ static void l2cap_chan_seg_sent(struct bt_conn *conn, void *user_data)
 	l2cap_chan_tx_resume(BT_L2CAP_LE_CHAN(chan));
 }
 
+static bool atomic_test_and_dec(atomic_t *target)
+{
+	static struct k_spinlock lock;
+	k_spinlock_key_t key;
+
+	key = k_spin_lock(&lock);
+
+	if (!target) {
+		k_spin_unlock(&lock, key);
+		return false;
+	}
+
+	target--;
+
+	k_spin_unlock(&lock, key);
+
+	return true;
+}
```

Comment:
`l2cap.c` doesn't seem the right place for this. Especially when it uses the reserved `atomic_*` namespace. I'd suggest to split this off into its own patch that adds it to `include/sys/atomic.h`

Reply:
I was a little hesitant to do that right away since it is not always easy to get things in the kernel side, but perhaps on the atomic.h that would be acceptable.

Reply:
Btw, the implementation doesn't look correct since it's looking at and decrementing the pointer rather than the value behind it

Reply:
I experiement a little bit with it residing on atomic.h, the problem is that there is no built-in for it so it needs to be inline and add includes for spinlock, so I don't think it is very clean:

https://gist.github.com/Vudentz/a6b7a1cf93d4e984641107293f6ee900

Reply:
> I experiement a little bit with it residing on atomic.h, the problem is that there is no built-in for it so it needs to be inline and add includes for spinlock, so I don't think it is very clean

I've added @andrewboie and @andyross as reviewers (marked as codeowners for atomic.h) who can hopefully provide some guidance here. Why do you need to have duplicate implementations in the header and c-file? Why can't you just have a prototype in the header and the implementation in the c-file (which could then use the same lock as other functions there)?

Reply:
@jhedberg so there's three options for atomic ops, selected in a menu in kernel/Kconfig:

- CONFIG_ATOMIC_OPERATIONS_BUILTIN - This uses the compiler's builtin functions for atomic operations. This is the preferred option, but not all compilers have builtins so we have to fallback on some other options.
- CONFIG_ATOMIC_OPERATIONS_CUSTOM - The arch code has custom assembly routines that Zephyr maintains for these functions.
- CONFIG_ATOMIC_OPERATIONS_C - The worst option. These grab a spinlock and then just do the math non-atomically. Because spinlocks are reserved for kernel use, user mode has to make syscalls. Pretty much for CPUs that lack native atomic instructions.

If you look at the #ifdefery in the header you will see how these implementations are managed, the builtin option has these as simple inlines which translate to the GCC atomics.

Reply:
BTW https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html GCC does not have a builtin with "test and dec" semantics so I urge you guys to consider re-doing your code to use other atomic ops to get what you need.